core/drand_beacon_public.go

```diff
@@ -21,7 +22,13 @@ func (bp *BeaconProcess) BroadcastDKG(c context.Context, in *drand.DKGPacket) (*
 	addr := net.RemoteAddress(c)
 
 	if bp.dkgInfo == nil {
-		return nil, fmt.Errorf("drand: no dkg running and yet received a DKGPacket for beacon %s from node %s", bp.beaconID, addr)
+		bp.state.Unlock()
```

Comment:
should this lock and unlock be the other way around?

Reply:
Read the code :P

Reply:
But won’t locking before rechecking `dkgInfo` (without unlocking after) potentially deadlock the beacon process? 

Checking `dkgInfo` twice is also kinda funky - in the DKG refactor I’ve just set a grace period to avoid this

Reply:
The Unlock for the second lock is provided by the `defer bp.state.Unlock()` a few lines above.
I agree it's not the easiest to read code. Maybe we can refactor this in the future to remove the need to do the unlock/lock dance?

Reply:
Refactor will happen anyway in the DKG work, I hope this race condition in the DKG will entirely disappear, or that we'll store incoming DKG packet in some kind of buffer until DKG starts for our node, or something like that.

Reply:
FWIW, this is really just a hack to make CI tests pass more reliably without dropping DKG packets because one node hasn't fully initialized its DKG yet and yet receives broadcasts